### PR TITLE
Cross-compile to Scala 3

### LINKED ISSUE #90: Cross build to Scala 3
Are there any plans to target Scala 3? Is there anything standing in the way currently? I would be willing to help with the refactor if there is interest.

### PR DESCRIPTION
**Pull request by ioleo**
This PR sets up cross-compilation to Scala 3.

Closes #90.

### CODE REVIEW

File: project/BuildHelper.scala [hunk 1]
```diff
@@ -4,16 +4,17 @@ import scalafix.sbt.ScalafixPlugin.autoImport._
 
 object BuildHelper {
 
-  val Scala212: String = "2.12.17"
-  val Scala213: String = "2.13.10"
+  val Scala212: String   = "2.12.17"
+  val Scala213: String   = "2.13.10"
+  val ScalaDotty: String = "3.2.2"
```
Comment (drmarjanovic) [inline project/BuildHelper.scala:9]: Can we use `Scala3` instead of `ScalaDotty`?

File: project/BuildHelper.scala [hunk 2]
```diff
@@ -23,10 +24,53 @@ object BuildHelper {
       autoAPIMappings := true
     )
 
-  private val extraOptions =
-    List("-Ywarn-unused")
+  private val stdOptions = Seq(
```
Comment (drmarjanovic) [inline project/BuildHelper.scala:27]: We prefer using `List` instead of `Seq`.

File: project/BuildHelper.scala [hunk 3]
```diff
@@ -23,10 +24,53 @@ object BuildHelper {
...
+    "-Yrangepos",
+    "-Xlint:_,-missing-interpolator,-type-parameter-shadow",
+    "-Ywarn-numeric-widen",
+    "-Ywarn-value-discard"
+  )
 
+  def extraOptions(scalaVersion: String) =
+    CrossVersion.partialVersion(scalaVersion) match {
+      case Some((3, 2)) =>
+        Seq(
```
Comment (drmarjanovic) [inline project/BuildHelper.scala:48]: We should use `List` instead of `Seq` everywhere.

File: project/BuildHelper.scala [hunk 4]
```diff
@@ -23,10 +24,53 @@ object BuildHelper {
...
+          "-Ywarn-infer-any",
+          "-Ywarn-nullary-override",
+          "-Ywarn-nullary-unit",
+          "-Ywarn-unused:params,-implicits",
+          "-Xfuture",
+          "-Xsource:2.13",
+          "-Xmax-classfile-name",
+          "242"
+        )
+      case _ => Seq()
```
Comment (drmarjanovic) [inline project/BuildHelper.scala:74]: Can we use `Nil` instead of `Seq()`?

**Review by drmarjanovic — COMMENTED** [review]
Hi @ioleo, great work overall. I'm thrilled because we are adding Scala 3 support!

Can you add Scala 3 to our [ci.yml](https://github.com/lambdaworks/zio-elasticsearch/blob/main/.github/workflows/ci.yml) to verify it through tests as well?

Once again, thank you and we are looking forward to more contributions from you!

**Review by drmarjanovic — COMMENTED** [review]
@ioleo, thanks for the quick response!

Can you just run `sbt prepare` before pushing? Seems our linter is failing! Thanks 🙏🏻

**Comment by ioleo** [comment]
Hmm, I don't get the error here. `assert` and `assertCustom` are members of `NewType[String]`, so why would Scala 3 not find those?

**Comment by ioleo** [comment]
@drmarjanovic ZIO Prelude 3.x is missing `assert` and `assertCustom` as they were backed by Scala 2 macro.
https://github.com/zio/zio-prelude/blob/series/2.x/core/shared/src/main/scala-2/zio/prelude/Newtype.scala#L258

Can we replace those calls (to `assert` and `assertCustom`)? Maybe more boilerplatish, but not relying on macros?
I haven't used ZIO Prelude, so I'm not sure what these do and why do you need them here in the codebase.

**Comment by ioleo** [comment]
Continued in #131

**Pull request closed without merging**